zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobVariablesCollector.java

```diff
@@ -7,41 +7,87 @@
-    if (elementInstanceKey >= 0) {
-      final DirectBuffer variables = collectVariables(requestedVariables, elementInstanceKey);
-      jobRecord.setVariables(variables);
+    final DirectBuffer processVariables;
+    if (elementInstanceKey < 0) {
+      processVariables = DocumentValue.EMPTY_DOCUMENT;
+    } else if (requestedVariables.isEmpty()) {
+      processVariables = variableState.getVariablesAsDocument(elementInstanceKey);
     } else {
-      jobRecord.setVariables(DocumentValue.EMPTY_DOCUMENT);
+      processVariables =
+          variableState.getVariablesAsDocument(elementInstanceKey, requestedVariables);
     }
-  }
 
-  private DirectBuffer collectVariables(
-      final Collection<DirectBuffer> variableNames, final long elementInstanceKey) {
-    final DirectBuffer variables;
+    final DirectBuffer jobVariables =
+        switch (jobRecord.getJobKind()) {
+          case BPMN_ELEMENT, EXECUTION_LISTENER -> processVariables;
+          case TASK_LISTENER -> {
+            final var taskVariablesMap = getTaskVariables(requestedVariables, elementInstanceKey);
 
-    if (variableNames.isEmpty()) {
-      variables = variableState.getVariablesAsDocument(elementInstanceKey);
-    } else {
-      variables = variableState.getVariablesAsDocument(elementInstanceKey, variableNames);
-    }
+            // merge the two variables maps favoring the task variables over process variables
+            final Map<String, Object> taskListenersVariables =
+                MsgPackConverter.convertToMap(processVariables);
+            taskListenersVariables.putAll(taskVariablesMap);
+            yield BufferUtil.wrapArray(MsgPackConverter.convertToMsgPack(taskListenersVariables));
+          }
+        };
+
+    jobRecord.setVariables(jobVariables);
+  }
 
-    return variables;
+  private Map<String, Object> getTaskVariables(
+      final Collection<DirectBuffer> requestedVariables, final long elementInstanceKey) {
+    final var elementInstance = elementInstanceState.getInstance(elementInstanceKey);
+    if (elementInstance == null) {
+      return Map.of();
+    }
+    final var userTaskIntermediateState =
+        userTaskState.getIntermediateState(elementInstance.getUserTaskKey());
+    if (userTaskIntermediateState == null) {
+      return Map.of();
+    }
+    final var taskVariables = userTaskIntermediateState.getRecord().getVariablesBuffer();
+    if (taskVariables.capacity() <= 0) {
+      return Map.of();
+    }
+    final Map<String, Object> taskVariablesMap = MsgPackConverter.convertToMap(taskVariables);
+    if (requestedVariables.isEmpty()) {
+      return taskVariablesMap;
+    }
+    final Map<String, Object> filteredTaskVariablesMap = new HashMap<>();
+    taskVariablesMap.forEach(
+        (key, value) -> {
+          if (requestedVariables.contains(BufferUtil.wrapString(key))) {
+            filteredTaskVariablesMap.put(key, value);
+          }
+        });
+    return filteredTaskVariablesMap;
```

Comment:
🔧 We can streamline this method by removing the need for the `filteredTaskVariablesMap` variable and directly using the Streams API to filter `taskVariablesMap`:

```suggestion
    return taskVariablesMap.entrySet().stream()
        .filter(e -> requestedVariables.contains(BufferUtil.wrapString(e.getKey())))
        .collect(toMap(Entry::getKey, Entry::getValue));
```